das/worker.go

```diff
@@ -138,14 +101,80 @@ func (w *worker) run(
 	}
 }
 
-func newWorker(j job) worker {
-	return worker{
-		state: workerState{
-			job:    j,
-			Curr:   j.From,
-			failed: make([]uint64, 0),
-		},
+func (w *worker) sample(ctx context.Context, timeout time.Duration, height uint64) error {
+	h, err := w.getHeader(ctx, height)
+	if err != nil {
+		return err
 	}
+
+	start := time.Now()
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	err = w.sampleFn(ctx, h)
+	w.metrics.observeSample(ctx, h, time.Since(start), err, w.state.isRecentHeader)
+	if err != nil {
+		if !errors.Is(err, context.Canceled) {
+			log.Debugw(
+				"failed to sample header",
+				"height", h.Height(),
+				"hash", h.Hash(),
+				"square width", len(h.DAH.RowsRoots),
+				"data root", h.DAH.Hash(),
+				"err", err,
+				"finished (s)", time.Since(start),
+			)
+		}
+		return err
+	}
+
+	log.Debugw(
+		"sampled header",
+		"height", h.Height(),
+		"hash", h.Hash(),
+		"square width", len(h.DAH.RowsRoots),
+		"data root", h.DAH.Hash(),
+		"finished (s)", time.Since(start),
+	)
+
+	// notify network about availability of new block data (note: only full nodes can notify)
+	if w.state.isRecentHeader {
+		err = w.broadcast(ctx, h.DataHash.Bytes())
+		if err != nil {
+			log.Warn("failed to broadcast availability message",
+				"height", h.Height(), "hash", h.Hash(), "err", err)
+		}
+	}
+	return nil
+}
+
+func (w *worker) getHeader(ctx context.Context, height uint64) (*header.ExtendedHeader, error) {
+	if w.state.header != nil {
+		return w.state.header, nil
+	}
+
+	// TODO: get headers in batches
+	start := time.Now()
+	h, err := w.getter.GetByHeight(ctx, height)
```

Comment:
shouldn't we still implement some reasonable timeout here as well? Or is this no longer necessary with the other timeouts PR?

Reply:
Sometimes DASing goes much faster than syncing headers, e.g, blocks are empty. In such a case, there is nothing else DASer can do besides waiting for the header. If we do timeout, they will be hit sometimes; we would crap out with logs and start to wait again. Vlad put a timeout before and had failure logs when bootstrappers didn't serve new headers and after this, we decided to avoid timeout here. Again, because there is nothing we can do and should just wait.

Reply:
I would also like to add that if syncing is stuck for any reason, only catchup workers will be locked. With changes from this PR recent jobs will keep sampling recent headers

Reply:
ok